addons/udes_stock/tests/common.py

```diff
@@ -497,11 +511,25 @@ def create_partner(cls, name, **kwargs):
         return Partner.create(vals)
 
     @classmethod
-    def complete_picking(cls, picking, validate=True):
-        """Marks a picking and all its moves as done"""
+    def complete_picking(cls, picking, validate=True, set_dest_location=True):
+        """
+        Marks a picking and all its moves as done.
+
+        Validates the picking unless `validate` is set to False.
+
+        If the destination location on the move line isn't an internal location,
+        then it will be set to the first internal child of that location if possible
+        unless `set_dest_location` is set to False.
+        """
         for move_line in picking.move_line_ids:
             if move_line.state == "assigned":
                 move_line.qty_done = move_line.product_uom_qty
+                if move_line.location_dest_id.usage != "internal":
```

Comment:
I can see how this works well now that we have blocked down the location structure but is it still a good idea doing this since we are forcing a default behaviour in the tests that is not done in UDES?

Reply:
I'm not sure why you mean by tests not done in UDES. Could you give me an example?

The reason this was needed was because previously view locations were being turned into internal locations, I assume to save time. Without this step the tests that use this helper method will error if using the default location of a picking type, which will almost always be a view location.

Reply:
What I'm trying to say is that this helper method is making automatic a "process" than in UDES doesn't exist.

For example, let's say that we have a picking where the move lines have the done quantities equal to the requested quantities (hence they are done) but the destination location is not set to anything different that the default destination location of the picking (which is usually a location of type view), then when we try to validate the picking UDES will complain saying "You cannot take products from or deliver products to a location of type 'view'". This is because UDES does not automatically set the destination location to the first child of the default destination location. 

So, here in the tests it is not wrong to do anything like this which UDES doesn't do for you but effectively this routine is doing that automatically for all the tests. It's not wrong since it's an helper function and it's job is to make our life easier but I feel like we are hiding this (once explicit behaviour) from the code of each individual test. Again, it's not a real issue but more of a consideration that I'm trying to make.

Reply:
It also validates the picking automatically, unless `validate` is set to False.

The point of the helper method is to speed up tests. For most tests the specific destination location isn't really important, so this is just saving time.

If it was important, the dev has the same option they have now. Manually setting the destination location on the move line(s) and then calling the helper method to complete the picking (because it only automatically selects a child if the type isn't internal already). And of course they can just set `set_dest_location` to False.

At the end of the day the helper methods are there to make our lives easier in terms of tests like you said. It's still the responsibility of the dev to make sure they understand what the helper methods are doing and whether it is appropriate to use it.

Reply:
Make sense, if that's the case where most tests don't require this level of "hardening" on the requirements then I'm happy with this helper function and I'm happy that we discussed it.